crates/ruff/src/args.rs

```diff
@@ -622,6 +745,154 @@ fn resolve_bool_arg(yes: bool, no: bool) -> Option<bool> {
+}
+
+impl std::fmt::Display for TomlParseFailure {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let TomlParseFailure {
+            kind,
+            underlying_error,
+        } = self;
+        let display = format!("{kind}:\n\n{underlying_error}");
+        write!(f, "{}", display.trim_end())
+    }
+}
+
+/// Enumeration to represent a single `--config` argument
+/// passed via the CLI.
+///
+/// Using the `--config` flag, users may pass 0 or 1 paths
+/// to configuration files and an arbitrary number of
+/// "inline TOML" overrides for specific settings.
+///
+/// For example:
+///
+/// ```sh
+/// ruff check --config "path/to/ruff.toml" --config "extend-select=['E501', 'F841']" --config "lint.per-file-ignores = {'some_file.py' = ['F841']}"
+/// ```
+#[derive(Clone, Debug)]
+pub enum SingleConfigArgument {
+    FilePath(PathBuf),
+    SettingsOverride(Arc<Options>),
+}
+
+#[derive(Clone)]
+pub struct ConfigArgumentParser;
+
+impl ValueParserFactory for SingleConfigArgument {
+    type Parser = ConfigArgumentParser;
+
+    fn value_parser() -> Self::Parser {
+        ConfigArgumentParser
+    }
+}
+
+impl TypedValueParser for ConfigArgumentParser {
+    type Value = SingleConfigArgument;
+
+    fn parse_ref(
+        &self,
+        cmd: &clap::Command,
+        arg: Option<&clap::Arg>,
+        value: &std::ffi::OsStr,
+    ) -> Result<Self::Value, clap::Error> {
+        let path_to_config_file = PathBuf::from(value);
+        if path_to_config_file.exists() {
+            return Ok(SingleConfigArgument::FilePath(path_to_config_file));
+        }
+
+        let value = value
+            .to_str()
+            .ok_or_else(|| clap::Error::new(clap::error::ErrorKind::InvalidUtf8))?;
+
+        let toml_parse_error = match toml::Table::from_str(value) {
+            Ok(table) => match table.try_into() {
+                Ok(option) => return Ok(SingleConfigArgument::SettingsOverride(Arc::new(option))),
+                Err(underlying_error) => TomlParseFailure {
+                    kind: TomlParseFailureKind::UnknownOption,
+                    underlying_error,
+                },
+            },
+            Err(underlying_error) => TomlParseFailure {
+                kind: TomlParseFailureKind::SyntaxError,
+                underlying_error,
+            },
+        };
+
+        let mut new_error = clap::Error::new(clap::error::ErrorKind::ValueValidation).with_cmd(cmd);
+        if let Some(arg) = arg {
+            new_error.insert(
+                clap::error::ContextKind::InvalidArg,
+                clap::error::ContextValue::String(arg.to_string()),
+            );
+        }
+        new_error.insert(
+            clap::error::ContextKind::InvalidValue,
+            clap::error::ContextValue::String(value.to_string()),
+        );
+
+        // small hack so that multiline tips
+        // have the same indent on the left-hand side:
+        let tip_indent = " ".repeat("  tip: ".len());
```

Comment:
A little concerning how much implicit coupling there is to Clap, but I don't have better suggestions (and at least there's test coverage, I think, such that if we upgrade Clap and the output format changes, we will know that there's a disconnect here).

Reply:
Yeah. Ideally clap would have a feature that allows long error messages to be wrapped to the terminal width, much as it does with the output of `--help`. But it doesn't seem to have that feature right now :/

Initially I had each sentence as a separate tip, which obviated the need for this kind of hack, but Zanie (correctly!) pointed out that that looked pretty weird